### PR TITLE
Also read default custom CMOR tables if custom location is specified

### LINKED ISSUE #2278: If custom location for custom CMOR tables is used, default location is ignored
**Describe the bug**
If a custom location for custom CMOR tables is used via `custom: cmor_path: ...` in `config-developer`, the default location `esmvalcore/cmor/tables/custom` is ignored. To avoid copying all existing tables to a custom location, this should be fixed.

I will open a PR for this shortly.

@FranziskaWinterstein

### PR DESCRIPTION
**Pull request by schlunma**
## Description

This PR makes sure that the default custom CMOR tables are always read, even when a custom location is specified.

Closes #2278

Link to documentation: https://esmvaltool--2279.org.readthedocs.build/projects/ESMValCore/en/2279/quickstart/configure.html#custom-cmor-tables

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] ~All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful~ (Codacy issues were already present in main).

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by schlunma** [comment]
@FranziskaWinterstein could you try if this solves your problem with the custom tables? Thanks!!

File: doc/quickstart/configure.rst
Comment (valeriupredoi) [inline doc/quickstart/configure.rst:702]:
```suggestion
For derived variables (the ones with ``derive: true`` in the recipe), the custom CMOR tables
```

File: doc/quickstart/configure.rst
Comment (valeriupredoi) [inline doc/quickstart/configure.rst:711]: is that a good idea? I'd argue CMOR stock should be the first port of entry

File: esmvalcore/cmor/table.py
Comment (valeriupredoi) [inline esmvalcore/cmor/table.py:1020]: this is slightly confusing variable naming (am not pointing fingers at you, since these were already in, but maybe a good chance to change them?) default CMOR stock: cmor tables, custom tables inside ESMValCore: custom tables, user specified tables: user tables IMHO

**Review by valeriupredoi — CHANGES_REQUESTED** [review]
many thanks, Manu! A couple very minor points from me :beer:

File: doc/quickstart/configure.rst
Comment (schlunma) [reply doc/quickstart/configure.rst:711]: I think it is; it gives the user the option change an existing custom variable. For example, this might be necessary if a native model uses a variable name that we also have in our custom CMOR tables, but e.g. with different dimensions.

Comment (valeriupredoi) [reply doc/quickstart/configure.rst:711]: indeed - I can see the benefits, but I can also see the negatives (stuff that can not be reproduced anymore bc the user has left with their tables). At any rate, it's philosophical at this point, the feature's in and this PR improves it :+1:

File: esmvalcore/cmor/table.py
Comment (valeriupredoi) [reply esmvalcore/cmor/table.py:1020]: prob best addressed in a separate PR BTW

**Review by valeriupredoi — APPROVED** [review]

File: esmvalcore/cmor/table.py
Comment (schlunma) [reply esmvalcore/cmor/table.py]: Agreed! I changed the new `_custom_cmor_folder` to `_user_table_dir` in https://github.com/ESMValGroup/ESMValCore/pull/2279/commits/feed776791ad0bce8e08c2bd44048e8908409167, but did not change `_cmor_folder` for consistency with the other tables 👍

**Comment by valeriupredoi** [comment]
BTW @schlunma I think this closes this issue too https://github.com/ESMValGroup/ESMValCore/issues/1388 - may be worth keeping that tool issue open until this gets merged and either an RC or full new Core version is out and about :beer:

**Comment by schlunma** [comment]
> BTW @schlunma I think this closes this issue too #1388 🍺

Nah I don't think so, this PR only changes the behavior if a custom location for the custom location is specified, #1388 also appears with our default location for the custom tables (I think 😁).

**Comment by valeriupredoi** [comment]
ah well, worth the try, am in Xmas cleaning mode :grin:

**Comment by FranziskaWinterstein** [comment]
@schlunma Thank you for taking care of this. However, I have difficulties in executing my recipe. I was able to reproduce the former error, but with your changes the following happens: `ERROR:esmvalcore._main:Program terminated abnormally, see stack trace below for more information:
Traceback (most recent call last):
  File "/home/b/b309109/ESMValCore/esmvalcore/_main.py", line 518, in run
    fire.Fire(ESMValTool())
  File "/home/b/b309109/ESMValCore/esmvalcore/_main.py", line 319, in __init__
    self.__setattr__(entry_point.name, entry_point.load()())
  File "/work/bd1132/b309109/conda_envs/envs/esmvaltool/lib/python3.10/importlib/metadata/__init__.py", line 171, in load
    module = import_module(match.group('module'))
  File "/work/bd1132/b309109/conda_envs/envs/esmvaltool/lib/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/b/b309109/ESMValTool/esmvaltool/cmorizers/data/cmorizer.py", line 17, in <module>
    from esmvalcore._task import write_ncl_settings
  File "/home/b/b309109/ESMValCore/esmvalcore/_task.py", line 24, in <module>
    from ._citation import _write_citation_files
  File "/home/b/b309109/ESMValCore/esmvalcore/_citation.py", line 10, in <module>
    from .config._diagnostics import DIAGNOSTICS
  File "/home/b/b309109/ESMValCore/esmvalcore/config/__init__.py", line 11, in <module>
    from ._config_object import CFG, Config, Session
  File "/home/b/b309109/ESMValCore/esmvalcore/config/_config_object.py", line 242, in <module>
    CFG = Config._load_user_config(USER_CONFIG, raise_exception=False)
  File "/home/b/b309109/ESMValCore/esmvalcore/config/_config_object.py", line 70, in _load_user_config
    new.update(mapping)
  File "/work/bd1132/b309109/conda_envs/envs/esmvaltool/lib/python3.10/_collections_abc.py", line 999, in update
    self[key] = other[key]
  File "/home/b/b309109/ESMValCore/esmvalcore/config/_validated_config.py", line 65, in __setitem__
    raise InvalidConfigParameter(
esmvalcore.exceptions.InvalidConfigParameter: `offline` is not a valid config parameter.
`

**Comment by schlunma** [comment]
It looks like you are using the removed option `offline` in one of your config files (could also be the default one at `~/.esmvaltool/config-user.yml`). `offline` has been deprecated in v2.8 and removed just now for v2.10 (https://github.com/ESMValGroup/ESMValCore/pull/2213). The new option is called `search_esgf` (details [here](https://github.com/ESMValGroup/ESMValCore/pull/1935)).

**Comment by schlunma** [comment]
See https://github.com/ESMValGroup/ESMValCore/issues/2280

**Comment by FranziskaWinterstein** [comment]
Okay, after solving all my config issues, the behaviour concerning the custom tables is as I would expect it, i.e. favouring the external folder over the internal only if duplicate tables are defined. Thank you!

**Comment by valeriupredoi** [comment]
@schlunma - I approved this PR last year :grin: I can also merge if I get the nod of OK from you :+1:

**Comment by schlunma** [comment]
> @schlunma - I approved this PR last year 😁 I can also merge if I get the nod of OK from you 👍

Yeah, please merge :grin:

**Pull request merged**
